Add `excludefromcoinjoin` from RPC

turbolay opened the pull request:
I thought of adding an RPC endpoint to access #9925 feature, as the UI was never made.
The function can be used to both exclude some coins or "unexclude" them, depending on the bool passed.
I can change the naming and signature if anyone has a better idea. I can also move the business logic.

I also took the liberty to implement 2 ways to exclude coins: By Outpoint or by Label of a transaction linked to the coin.
I can remove the "By Label" feature if it's preferred to don't have it.

adamPetho commented:
> would be good to also add an rpc method to undo the "exclusion".

The function can be used to both exclude some coins or "unexclude" them, depending on the `boolen` passed.

MarnixCroes commented:
> > would be good to also add an rpc method to undo the "exclusion".
> 
> The function can be used to both exclude some coins or "unexclude" them, depending on the `boolen` passed.

ah, should've checked first.
thanks for pointing out

lontivero commented on WalletWasabi.Daemon/Rpc/WasabiJsonRpcService.cs:
```diff
@@ -253,6 +253,47 @@ public object[] GetHistory()
 	}
 
+	[JsonRpcMethod("excludefromcoinjoin")]
+	public IEnumerable<OutPoint> ExcludeFromCoinjoin(bool exclude, string? password = null, OutPoint[]? outPoints = null, string[]? labels = null)
```
Please design the API some how, this is not good and I am pretty sure it doesn't work. Have you tested it?

It seems to me this method tries to do too much, in fact it tries to do two things:
* exclude by outpoint. OK
* exclude by labels. Not OK. RPC is an interface to build things. 

Anyway, even if exclude by labels is desired, the interface would be really complex so, it would be better to have another method.

turbolay replied:
It was working; just I forgot to call the update. Anyway, I made this draft PR exactly to have this kind of feedback about the signature/implementation, so thanks.

lontivero commented on WalletWasabi.Daemon/Rpc/WasabiJsonRpcService.cs:
```diff
@@ -253,6 +253,47 @@ public object[] GetHistory()
+		AssertWalletIsLoggedIn(activeWallet, password ?? "");
+
+		var updatedCoinsOutPoints = new List<OutPoint>();
+		foreach (var coin in activeWallet.GetAllCoins())
```
RPC methods are just another interface to business logic. There should not be logic here (loops, control flow statements) except to build the outputs.

turbolay replied:
I moved it to Wallet

lontivero reviewed (changes requested):
Concept ACK. Implementation NACK

turbolay commented:
I addressed reviews (moved logic to `Wallet` class and removed code about `Labels`, I will add this in another PR).
I still have a doubt about what should be the order of the optional parameters `exclude` and `password`, and if it would be better to have a new function instead of the optional `exclude` parameter.

lontivero commented:
How to test:

* Run the `daemon`
```
$ dotnet run --usetor=false --jsonrpcserverenabled=true --wallet=<wallet-name> --network=testnet
```

* List the unspent coins and select one (the latest one in the following case)
```
$ ./wcli.sh -wallet=<wallet-name> listunspentcoins | tail -1

e060ca31d3d3743a45c507e6dc09ebac7c7d6974da16e5219bb91a404e9b4856  0      100000  1                   true       4292535561     84'/0'/0'/1/1511  tb1qvskwhrg4jhkv5peq2sunqc4c2pcahlrv5vc5k8                      false 
```

* Exclude it from coinjoins
```
./wcli.sh -wallet=<wallet-name> excludefromcoinjoin e060ca31d3d3743a45c507e6dc09ebac7c7d6974da16e5219bb91a404e9b4856 0
```

* Check it is excluded (see the latest column)

```
$ ./wcli.sh -wallet=<wallet-name> listunspentcoins | tail -1

e060ca31d3d3743a45c507e6dc09ebac7c7d6974da16e5219bb91a404e9b4856  0      100000  1                   true       4292535561     84'/0'/0'/1/1511  tb1qvskwhrg4jhkv5peq2sunqc4c2pcahlrv5vc5k8                      true 
```

* Revert it by remove the exclusion
```
./wcli.sh -wallet=<wallet-name> excludefromcoinjoin e060ca31d3d3743a45c507e6dc09ebac7c7d6974da16e5219bb91a404e9b4856 0 false
```

Check it again.
